clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -1135,16 +1137,42 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_traced_schedule()
     res = remove_zero_instruction_segments(input_sched, all_sched);
     if (res != sched_type_t::STATUS_SUCCESS)
         return res;
-    for (int output_idx = 0; output_idx < static_cast<output_ordinal_t>(outputs_.size());
-         ++output_idx) {
+    // Sort by cpuid to get a more natural ordering.
+    // Probably raw2trace should do this in the first place, but we have many
+    // schedule files already out there so we still need a sort here.
+    std::sort(disk2index.begin(), disk2index.end(),
+              [disk2cpuid](const output_ordinal_t &l, const output_ordinal_t &r) {
+                  return disk2cpuid[l] < disk2cpuid[r];
+              });
+    // disk2index now holds the sorted index, but we need another array to
+    // store that in the disk order. E.g., if the original file held 6,2,3,7,
+    // disk2index would then hold 1,2,0,3, but we want 2,0,1,3.
```

Comment:
Still trying to wrap my head around this. Maybe I need to look more closely.

We have disk2output disk2index and disk2cpuid. Do all of them have the "ordinal of schedule entry on disk" as the index? If not, the identifiers could be made more descriptive to differentiate between what the index is for the respective array.

Reply:
It took me longer than I'd like to admit to completely follow this. Some suggestions.

Add narration:

disk2output[i] stores the output_idx for the cpu at ith ordinal in the cpu_schedule_file.
disk2cpuid[i] stores the actual cpu_id for the cpu at ith ordinal in the cpu_schedule_file
disk2index[i] initially stores i, but eventually (after the sorting step) stores the ordinal in the cpu_schedule_file that has the ith largest cpuid.

Maybe rename disk2index. Eventually (after the sorting step) the key isn't the ordinal in the cpu_schedule_file anymore.

Maybe combine disk2cpuid and disk2index into one struct.

Clarify what "disk" means. It means "ordinal in the cpu schedule file on disk". Maybe more verbose naming would be better. Maybe sched_idx is better.

Reply:
It also took me a little while to figure out how to do this as well so agreed it may need more explaining.  If there weren't cross-indices pointing into all_sched I would have just sorted that (added a comment about that).

I changed s/disk/disk_ord/ and added a comment that it is the on-disk order.

Added descriptions of the vectors as suggested and elaborated the original comment a bit.